### ISSUE TITLE
Add optional context to `resolveCallableDependencies`

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#15080]
Now that modules have tree traversal we should add context to `resolveCallableDependencies`.

### What steps will reproduce the problem?

Application has `User` component.
Module A has `User` component.

Somewhere in Module A try to use `resolveCallableDependencies` on something that requires a `User` component.

### What is the DESIRED result?

I'd love to get my modules' `User` component.

### What do you get instead?
The applications' `User` component.

### Additional info

Basically the way `\Yii::$app` is used in the container is as a preferred `ServiceLocator`. Instead we could support passing in any `ServiceLocator` as a context.

Current signature:

```php
public function resolveCallableDependencies(callable $callback, $params = [])
```
Proposed signature:

```php
public function resolveCallableDependencies(callable $callback, $params = [], ServiceLocator $context = null)
```

Temporary backwards compatible implementation:

```php
public function resolveCallableDependencies(callable $callback, $params = [], ServiceLocator $context = null) {
if (!isset($context)) {
    $context = \Yii::$app;
}
```
At some point we should then remove the default and always require explicit context (thus removing the access to the global).

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.13
| PHP version      | 
| Operating system |

**Comment by SamMousa** [comment]
Any input, i'm happy to defend the proposal if anyone has an opinion :)

**Comment by samdark** [comment]
It makes sense to me. @yiisoft/core-developers what do you think?

**Comment by dynasource** [comment]
agree. It makes sense to make `Yii::$app` in `resolveCallableDependencies` customizable

**Comment by samdark** [comment]
Can't do so w/o breaking BC.

**Comment by SamMousa** [comment]
Why not? Adding an optional argument is allowed right?

Anyway, I'm fine with pushing this to 2.1

**Comment by samdark** [comment]
> Why not? Adding an optional argument is allowed right?

**Comment by SamMousa** [comment]
@samdark for which version should I make a PR?

Or should this be closed since you're working on a new DI project that will change everything?

**Comment by samdark** [comment]
I'm still unsure if we'll be able to use that DI library in 3.0.0 since it may change framework too much. Will get an answer in a few weeks (hopefully).

**Comment by SamMousa** [comment]
So definitely not in 2.x. Should I make a PR for 2.0.x then?

**Comment by SamMousa** [comment]
Ah, but won't there be a 2.1? Or we going to 3 immediately?
